### PR TITLE
storage: Deny localhost/private-ips in storage connections

### PR DESCRIPTION
**Pull request by rjobanp**
### Motivation

  * This PR fixes a recognized bug. Fixes https://github.com/MaterializeInc/console/issues/1635

### Tips for reviewer

There were several approaches discussed to implement this security fix, but we felt that the simplest was to simply resolve DNS ourselves for all storage connections and validate that the resolved address is not to a local or private-ip, thereby preventing a malicious user from attempting to create a Materialize Source that connects to some service on the materialize network. Most of these attacks would already be prevented by Cilium at the K8s networking layer, but this adds additional security.

This feature is activated by a LD feature flag, which I plan to use for testing once this goes to Staging, since it's difficult to validate that it actually allows valid connections in local testing. Instead the tests just validate that when activated, it prevents local/private-ip connections.

This also fixes some bugs in the dyncfg implementation for storage #26076 which was accidentally dropping the dyncfg values when performing a StorageConfiguration update.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage/src/storage_state.rs
Comment (rjobanp) [inline src/storage/src/storage_state.rs:1205]: @guswynn this bug prevented the `ALTER SYSTEM SET storage_enforce_external_addresses = true` from working. I think we were accidentally taking the option value from the updates which meant it didn't get propagated  to the `InternalStorageCommand::UpdateConfiguration` command

Comment (guswynn) [reply src/storage/src/storage_state.rs:1205]: Can you break this out in its own pr please? This is a affects a critical parameter (`storage_dataflow_delay_sources_past_rehydration`) and we need to merge this fix before the release cut today

Sorry about this!

Comment (rjobanp) [reply src/storage/src/storage_state.rs:1205]: yes can do!

Comment (rjobanp) [reply src/storage/src/storage_state.rs:1205]: https://github.com/MaterializeInc/materialize/pull/26189

**Review by guswynn — COMMENTED** [review]
This looks good but there are a few things I think need to be fixed:
- ssh tunnels also do resolution and must be checked
- for completeness, though its probably not actually possible to spoof, we should check what aws privatelink hosts resolve to
- we can't unwrap in `rewrite_broker_addr`, we have to fallback to an known-unresolvable host (like in ssh failures) and communicate this error through a method like `tunnel_status`

**Comment by rjobanp** [comment]
> ssh tunnels also do resolution and must be checked

@guswynn I'm pretty sure I added coverage of all the places where we connect to an SSH bastion host. We don't want to do the resolution on the destination URL in the case of SSH tunnels, since customers may want to connect to a private-address within their network that is resolvable via the SSH bastion-host. We just need to ensure the bastion host itself is global to us.

**Comment by guswynn** [comment]
@rjobanp you are totally correct, sorry about that! The kafka error comment should be it then!

**Comment by rjobanp** [comment]
@guswynn fixed the failing test cases and addressed your feedback, mostly.

We do now avoid unwraping in the `rewrite_broker_addr` method and instead return a known unresolvable host if we can't resolve the broker address. We don't need to report the error like we do for SSH since we don't need to distinguish the 'namespace' of the error, and the error will still get propagated as a DNS error by the rdkafka itself when it tries to resolve the broker address.

I also opted not to do anything for privatelink connections -- there is no user input involved with getting the vpc-endpoint address and it seems pointless to complicate that logic which is already somewhat onerous to understand.

**Comment by guswynn** [comment]
@rjobanp I would rather we actually communicate a meaningful error to the user, instead of having them interpret whatever rdkafka says about the address `failed-broker-address.dev.materialize.com`, but if you want to add a `TODO` and leave that for later, that fine for me!

**Review by benesch — COMMENTED** [review]
The basic structure here looks good, but I have a wrench to throw into the works. Today, a user can specify a hostname that resolves to multiple IPs, and Materialize will try each of those IPs in order. After this change, Materialize will try only the first. That's a regression with possible availability implications. I know at least one of our customers uses DNS to provide HA for SSH tunnels, by round-robining traffic between two SSH bastion hosts.

It's going to require a good deal more elbow grease, but I think it's fundamentally possible to support this with the same structure you have here. We'll need each library (MySQL, PostgreSQL, Kafka, SSH tunnels) to support taking multiple addresses as input for the same hostname. Then `resolve_address` can return a `Vec<IpAddr>`, and we can pass that _list_ of IPs into each library.

Here's a patch to our PostgreSQL fork that integrates an upstream change to allow taking multiple IPs as input: https://github.com/MaterializeInc/rust-postgres/pull/22

The MySQL change is hopefully similarly straightforward, although we'll likely need to reintroduce our fork to avoid blocking on upstream merging our PR.

SSH is a little annoying, because we shell out to the `ssh` binary and there's no way AFIACT to pass in a _list_ of IPs, even though SSH does the right thing if you pass in a hostname that resolves to multiple IPs. But I think we can just adjust our SSH tunnel code to shell out to `ssh` multiple times, one per IP, and it won't be too bad.

rdkafka is going to be the most annoying. The underlying librdkafka API supports `resolve_cb`s that return multiple addresses, but it'll require some C goop to expose this functionality in rust-rdkafka. I can assist with this, if you'd like. We can also take the chance to allow rust-rdkafka's callback to be fallible, which should clean up the error handling a good bit.

**Comment by rjobanp** [comment]
Good point on this breaking support for HA DNS configs -- let me look into refactoring MySQL and the SSH connection stuff. I would appreciate the help on rust-rdkafka if you have time to look into that!

**Comment by rjobanp** [comment]
Okay I've reverted all the non-ssh kafka connection & kafka client changes so that we can get this PR merged without waiting for the support in our rdkafka fork that will allow providing multiple resolved IPs for broker connections. I've left a TODO in its place and will reword the attached issue after we merge. @guswynn ready for another look - thanks!

File: src/mysql-util/src/tunnel.rs
Comment (guswynn) [inline src/mysql-util/src/tunnel.rs:31]: can you document what `None` means? (here and for pg)

File: src/ore/src/netio/dns.rs
Comment (guswynn) [inline src/ore/src/netio/dns.rs:63]: These ones seem to cover more cases, so probably worth making a tracking issue in our repo for this! (actually let me know if you want help on the process of trying to push these through stabilization!)

File: src/postgres-util/src/tunnel.rs
Comment (guswynn) [inline src/postgres-util/src/tunnel.rs:239]:
```suggestion
                    // Associate each resolved ip with the exact same, singular host, for tls verification.
                    // We are required to do this dance because `tokio-postgres` enforces that the number of 'host' and 'hostaddr' values must be the same.
                    
```

File: src/postgres-util/src/tunnel.rs
Comment (guswynn) [inline src/postgres-util/src/tunnel.rs:310]: same comment here

File: src/ssh-util/src/tunnel.rs
Comment (guswynn) [inline src/ssh-util/src/tunnel.rs:69]: We use this as a key when sharing ssh tunnels through the same bastion. To ensure we don't over-connect, this should be a `BTreeSet` so its sorted. Actually, we should sort in the pg and mysql connection code too.

Note that there are weird cases now where:
- a host resolves to a different number of ip's on different workers
- different workers connect to different upstream ips for the pg/mysql/etc server, if they appear connectable at different times

Both cases result in additional ssh connections, which is not something we can work around, but might be worth a comment

**Review by guswynn — APPROVED** [review]
looks good! I am accepting to unblock you, but before merging `resolve_address` should return a `BTreeSet` so all uses try ip's in order!

File: src/ore/src/netio/dns.rs
Comment (rjobanp) [reply src/ore/src/netio/dns.rs:63]: good call - made a tracking issue https://github.com/MaterializeInc/materialize/issues/26649

File: src/ssh-util/src/tunnel.rs
Comment (rjobanp) [reply src/ssh-util/src/tunnel.rs:69]: good call out! Switching to a `BtreeSet` across all connection types and will note those edge cases in a comment

**Pull request merged**
